From d5b492d4e15f2d1a1e726925566f2f7f44dbc7c5 Mon Sep 17 00:00:00 2001
From: Bing Niu <bing.niu@intel.com>
Date: Sat, 22 Apr 2017 13:38:31 +0800
Subject: [PATCH 1/5] vgt class introduction and adaption

introduce vgt class for vgpu life cycle manage, export paramter to
user, new class w/ pc initialise and adaption low mem.

Signed-off-by: Bing Niu <bing.niu@intel.com>
diff --git a/hw/display/Makefile.objs b/hw/display/Makefile.objs
index a59254a..879e7b0 100644
--- a/hw/display/Makefile.objs
+++ b/hw/display/Makefile.objs
@@ -1,5 +1,8 @@
 common-obj-$(CONFIG_ADS7846) += ads7846.o
 common-obj-$(CONFIG_VGA_CIRRUS) += cirrus_vga.o
+ifeq ($(findstring y,$(CONFIG_XEN)),y)
+obj-y += vgt_vga.o
+endif
 common-obj-$(CONFIG_G364FB) += g364fb.o
 common-obj-$(CONFIG_JAZZ_LED) += jazz_led.o
 common-obj-$(CONFIG_PL110) += pl110.o
diff --git a/hw/display/vgt_vga.c b/hw/display/vgt_vga.c
new file mode 100644
index 0000000..79b7dd2
--- /dev/null
+++ b/hw/display/vgt_vga.c
@@ -0,0 +1,426 @@
+/*
+ * QEMU vGT/XenGT Legacy VGA support
+ *
+ * Copyright (c) 2003 Fabrice Bellard
+ * Copyright (c) Citrix Systems, Inc
+ * Copyright (c) Intel Corporation.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#include "qemu/osdep.h"
+#include "hw/hw.h"
+#include "ui/console.h"
+#include "hw/i386/pc.h"
+#include "hw/pci/pci.h"
+#include "hw/pci/pci_host.h"
+#include "hw/pci/pci_bridge.h"
+#include "hw/pci/pci_bus.h"
+#include "vga_int.h"
+#include "ui/pixel_ops.h"
+#include "qemu/timer.h"
+#include "hw/loader.h"
+#include "qemu/log.h"
+#include "sysemu/arch_init.h"
+#include "hw/xen/xen.h"
+
+//#define DEBUG_VGT
+
+#ifdef DEBUG_VGT
+#define DPRINTF(fmt, ...) \
+    do { fprintf(stderr, "vgt: " fmt, ## __VA_ARGS__); } while (0)
+#else
+#define DPRINTF(fmt, ...) \
+    do { } while (0)
+#endif
+
+typedef struct VGTHostDevice {
+    PCIHostDeviceAddress addr;
+    int config_fd;
+    uint16_t vendor_id;
+    uint16_t device_id;
+    uint8_t revision_id;
+    uint16_t class_dev;
+} VGTHostDevice;
+
+typedef struct VGTVGAState {
+    PCIDevice dev;
+    struct VGACommonState state;
+    int num_displays;
+    VGTHostDevice host_dev;
+    bool instance_created;
+} VGTVGAState;
+
+/* These are the default values */
+int vgt_low_gm_sz = 64; /* in MB */
+int vgt_high_gm_sz = 448; /* in MB */
+int vgt_fence_sz = 4;
+int vgt_primary = 1; /* -1 means "not specified */
+
+static int vgt_host_device_get(VGTHostDevice *dev);
+static void vgt_host_device_put(VGTHostDevice *dev);
+
+void vgt_bridge_pci_write(PCIDevice *dev,
+                          uint32_t address, uint32_t val, int len)
+{
+#if 0
+    VGTVGAState *o = DO_UPCAST(VGTVGAState, dev, dev);
+#endif
+
+    assert(dev->devfn == 0x00);
+
+//  fprintf("vGT Config Write: addr=%x len=%x val=%x\n", addr, len, val);
+
+    switch (address) {
+#if 0
+        case 0x58:        // PAVPC Offset
+            xen_host_pci_set_block(o->host_dev, addr, val, len);
+            break;
+#endif
+    }
+
+    i440fx_write_config(dev, address, val, len);
+}
+
+/*
+ *  Inform vGT driver to create a vGT instance
+ */
+static void create_vgt_instance(void)
+{
+    /* FIXME: this should be substituded as a environment variable */
+    const char *path = "/sys/kernel/vgt/control/create_vgt_instance";
+    FILE *vgt_file;
+    int err = 0;
+
+    qemu_log("vGT: %s: domid=%d, low_gm_sz=%dMB, high_gm_sz=%dMB, "
+        "fence_sz=%d, vgt_primary=%d\n", __func__, xen_domid,
+        vgt_low_gm_sz, vgt_high_gm_sz, vgt_fence_sz, vgt_primary);
+    if (vgt_low_gm_sz <= 0 || vgt_high_gm_sz <=0 ||
+		vgt_primary < -1 || vgt_primary > 1 ||
+        vgt_fence_sz <=0) {
+        qemu_log("vGT: %s failed: invalid parameters!\n", __func__);
+        abort();
+    }
+
+    if ((vgt_file = fopen(path, "w")) == NULL) {
+        err = errno;
+        qemu_log("vGT: open %s failed\n", path);
+    }
+    /* The format of the string is:
+     * domid,aperture_size,gm_size,fence_size. This means we want the vgt
+     * driver to create a vgt instanc for Domain domid with the required
+     * parameters. NOTE: aperture_size and gm_size are in MB.
+     */
+    if (!err && fprintf(vgt_file, "%d,%u,%u,%u,%d\n", xen_domid,
+        vgt_low_gm_sz, vgt_high_gm_sz, vgt_fence_sz, vgt_primary) < 0) {
+        err = errno;
+    }
+
+    if (!err && fclose(vgt_file) != 0) {
+        err = errno;
+    }
+
+    if (err) {
+        qemu_log("vGT: %s failed: errno=%d\n", __func__, err);
+        exit(-1);
+    }
+}
+
+/*
+ *  Inform vGT driver to close a vGT instance
+ */
+static void destroy_vgt_instance(void)
+{
+    const char *path = "/sys/kernel/vgt/control/create_vgt_instance";
+    FILE *vgt_file;
+    int err = 0;
+
+    if ((vgt_file = fopen(path, "w")) == NULL) {
+        error_report("vgt: error: open %s failed", path);
+        err = errno;
+    }
+
+    /* -domid means we want the vgt driver to free the vgt instance
+     * of Domain domid.
+     * */
+    if (!err && fprintf(vgt_file, "%d\n", -xen_domid) < 0) {
+        err = errno;
+    }
+
+    if (!err && fclose(vgt_file) != 0) {
+        err = errno;
+    }
+
+    if (err) {
+        qemu_log("vGT: %s: failed: errno=%d\n", __func__, err);
+        exit(-1);
+    }
+}
+
+static int pch_map_irq(PCIDevice *pci_dev, int irq_num)
+{
+    return irq_num;
+}
+
+static void vgt_pci_conf_init_from_host(PCIDevice *dev,
+        uint32_t addr, int len)
+{
+    int ret;
+
+    if (len > 4) {
+        error_report("WARNIGN: length %x too large for config addr %x, ignore init",
+                len, addr);
+        return;
+    }
+
+    VGTHostDevice host_dev = {
+        .addr.domain = 0,
+        .addr.bus = pci_bus_num(dev->bus),
+        .addr.slot = PCI_SLOT(dev->devfn),
+        .addr.function = PCI_FUNC(dev->devfn),
+    };
+
+    /* FIXME: need a better scheme to grab the root complex. This
+     * only for a single VM scenario.
+     */
+    vgt_host_device_get(&host_dev);
+    ret = pread(host_dev.config_fd, dev->config + addr, len, addr);
+    if (ret < len) {
+        error_report("%s, read config addr %x, len %d failed.", __func__, addr, len);
+        return;
+    }
+    vgt_host_device_put(&host_dev);
+}
+
+static int vgt_host_pci_get_byte(VGTHostDevice *host_dev,
+                                  uint32_t addr, uint8_t *p)
+{
+    int ret;
+    uint8_t buf;
+
+
+    vgt_host_device_get(host_dev);
+    ret = pread(host_dev->config_fd, &buf, 1, addr);
+    if (ret < 1) {
+        error_report("%s, failed.", __func__);
+        return ret;
+    }
+    vgt_host_device_put(host_dev);
+
+    *p = buf;
+    return ret;
+}
+
+static void vgt_host_bridge_cap_init(PCIDevice *dev)
+{
+    assert(dev->devfn == 0x00);
+    uint8_t cap_ptr = 0;
+
+    VGTHostDevice host_dev = {
+        .addr.domain = 0,
+        .addr.bus = 0,
+        .addr.slot = 0,
+        .addr.function = 0,
+    };
+
+    vgt_host_pci_get_byte(&host_dev, PCI_CAPABILITY_LIST, &cap_ptr);
+    while (cap_ptr !=0) {
+        vgt_pci_conf_init_from_host(dev, cap_ptr, 4); /* capability */
+        vgt_pci_conf_init_from_host(dev, cap_ptr + 4, 4); /* capability */
+        vgt_pci_conf_init_from_host(dev, cap_ptr + 8, 4); /* capability */
+        vgt_pci_conf_init_from_host(dev, cap_ptr + 12, 4); /* capability */
+        //XEN_PT_LOG(pci_dev, "Add vgt host bridge capability: offset=0x%x, cap=0x%x\n", cap_ptr,
+        //    pt_pci_host_read(0, PCI_SLOT(pci_dev->devfn), 0, cap_ptr, 1) & 0xFF );
+        vgt_host_pci_get_byte(&host_dev, cap_ptr + 1, &cap_ptr);
+    }
+}
+
+void vgt_bridge_pci_conf_init(PCIDevice *pci_dev)
+{
+    printf("vgt_bridge_pci_conf_init\n");
+    printf("vendor id: %x\n", *(uint16_t *)((char *)pci_dev->config + 0x00));
+    vgt_pci_conf_init_from_host(pci_dev, 0x00, 2); /* vendor id */
+    printf("vendor id: %x\n", *(uint16_t *)((char *)pci_dev->config + 0x00));
+    printf("device id: %x\n", *(uint16_t *)((char *)pci_dev->config + 0x02));
+    vgt_pci_conf_init_from_host(pci_dev, 0x02, 2); /* device id */
+    printf("device id: %x\n", *(uint16_t *)((char *)pci_dev->config + 0x02));
+    vgt_pci_conf_init_from_host(pci_dev, 0x06, 2); /* status */
+    vgt_pci_conf_init_from_host(pci_dev, 0x08, 2); /* revision id */
+    vgt_pci_conf_init_from_host(pci_dev, 0x34, 1); /* capability */
+    vgt_host_bridge_cap_init(pci_dev);
+    vgt_pci_conf_init_from_host(pci_dev, 0x50, 2); /* SNB: processor graphics control register */
+    vgt_pci_conf_init_from_host(pci_dev, 0x52, 2); /* processor graphics control register */
+}
+
+static void vgt_reset(DeviceState *dev)
+{
+}
+
+static void vgt_cleanupfn(PCIDevice *dev)
+{
+    VGTVGAState *d = DO_UPCAST(VGTVGAState, dev, dev);
+
+    if (d->instance_created) {
+        destroy_vgt_instance();
+    }
+}
+
+static int vgt_initfn(PCIDevice *dev)
+{
+    VGTVGAState *d = DO_UPCAST(VGTVGAState, dev, dev);
+
+    DPRINTF("vgt_initfn\n");
+    d->instance_created = TRUE;
+
+    create_vgt_instance();
+    return 0;
+}
+
+static int vgt_host_device_get(VGTHostDevice *dev)
+{
+    char name[PATH_MAX];
+    int ret;
+
+    snprintf(name, sizeof(name), "/sys/bus/pci/devices/%04x:%02x:%02x.%x/config",
+             dev->addr.domain, dev->addr.bus, dev->addr.slot, dev->addr.function);
+    dev->config_fd = open(name, O_RDONLY);
+    if (dev->config_fd == -1) {
+        error_report("vgt: open failed: %s\n", strerror(errno));
+        return -1;
+    }
+
+    ret = pread(dev->config_fd, &dev->vendor_id, sizeof(dev->vendor_id), PCI_VENDOR_ID);
+    if (ret < sizeof(dev->vendor_id)) {
+        goto error;
+    }
+    ret = pread(dev->config_fd, &dev->device_id, sizeof(dev->device_id), PCI_DEVICE_ID);
+    if (ret < sizeof(dev->device_id)) {
+        goto error;
+    }
+    ret = pread(dev->config_fd, &dev->revision_id, sizeof(dev->revision_id), PCI_REVISION_ID);
+    if (ret < sizeof(dev->revision_id)) {
+        goto error;
+    }
+    ret = pread(dev->config_fd, &dev->class_dev, sizeof(dev->class_dev), PCI_CLASS_DEVICE);
+    if (ret < sizeof(dev->class_dev)) {
+        goto error;
+    }
+    DPRINTF("vendor: 0x%hx, device: 0x%hx, revision: 0x%hhx\n",
+           dev->vendor_id, dev->device_id, dev->revision_id);
+
+    return 0;
+
+error:
+    ret = ret < 0 ? -errno : -EFAULT;
+    error_report("vgt: Failed to read device config space");
+    return ret;
+}
+
+static void vgt_host_device_put(VGTHostDevice *dev)
+{
+    if (dev->config_fd >= 0) {
+        close(dev->config_fd);
+        dev->config_fd = -1;
+    }
+}
+
+DeviceState *vgt_vga_init(PCIBus *pci_bus)
+{
+    PCIDevice *dev;
+    PCIBridge *br;
+    VGTHostDevice host_dev = {
+        .addr.domain = 0,
+        .addr.bus = 0,
+        .addr.slot = 0x1f,
+        .addr.function = 0,
+    };
+
+    if (vgt_host_device_get(&host_dev) < 0) {
+        error_report("vgt: error: failed to get host PCI device");
+        return NULL;
+    }
+
+    if (host_dev.vendor_id != PCI_VENDOR_ID_INTEL) {
+        vgt_host_device_put(&host_dev);
+        error_report("vgt: error: vgt-vga is only supported on Intel GPUs");
+        return NULL;
+    }
+
+    vgt_host_device_put(&host_dev);
+
+    dev = pci_create_multifunction(pci_bus, PCI_DEVFN(0x1f, 0), true,
+                                   "vgt-isa");
+    if (!dev) {
+        error_report("vgt: error: vgt-isa not available");
+        return NULL;
+    }
+
+    qdev_init_nofail(&dev->qdev);
+
+    pci_config_set_vendor_id(dev->config, host_dev.vendor_id);
+    pci_config_set_device_id(dev->config, host_dev.device_id);
+    pci_config_set_revision(dev->config, host_dev.revision_id);
+    pci_config_set_class(dev->config, host_dev.class_dev);
+    br = PCI_BRIDGE(dev);
+    pci_bridge_map_irq(br, "IGD Bridge",
+                       pch_map_irq);
+
+    printf("Create vgt ISA bridge successfully\n");
+
+    dev = pci_create_multifunction(pci_bus, PCI_DEVFN(0x2, 0), true,
+                                   "vgt-vga");
+    if (!dev) {
+        error_report("vgt: error: vgt-vga not available");
+        return NULL;
+    }
+
+    qdev_init_nofail(&dev->qdev);
+    printf("Create vgt VGA successfully\n");
+    return DEVICE(dev);
+}
+
+static void vgt_class_initfn(ObjectClass *klass, void *data)
+{
+    DeviceClass *dc = DEVICE_CLASS(klass);
+    PCIDeviceClass *ic = PCI_DEVICE_CLASS(klass);
+    ic->init = vgt_initfn;
+    dc->reset = vgt_reset;
+    ic->exit = vgt_cleanupfn;
+    dc->vmsd = &vmstate_vga_common;
+}
+
+static TypeInfo vgt_info = {
+    .name          = "vgt-vga",
+    .parent        = TYPE_PCI_DEVICE,
+    .instance_size = sizeof(VGTVGAState),
+    .class_init    = vgt_class_initfn,
+};
+
+static TypeInfo isa_info = {
+    .name          = "vgt-isa",
+    .parent        = TYPE_PCI_BRIDGE,
+    .instance_size = sizeof(PCIBridge),
+};
+
+static void vgt_register_types(void)
+{
+    type_register_static(&vgt_info);
+    type_register_static(&isa_info);
+}
+
+type_init(vgt_register_types)
diff --git a/hw/i386/pc_piix.c b/hw/i386/pc_piix.c
index a54a468..f405d13 100644
--- a/hw/i386/pc_piix.c
+++ b/hw/i386/pc_piix.c
@@ -226,7 +226,19 @@ static void pc_init1(MachineState *machine,
 
     pc_register_ferr_irq(pcms->gsi[13]);
 
-    pc_vga_init(isa_bus, pcmc->pci_enabled ? pci_bus : NULL);
+    /*
+     * Initialize XenGT hooks before normal VGA init. The
+     * ideal case is to have IGD presented as the primary
+     * graphics card in 00:02.0, and then have other emulated
+     * PCI VGA card all disabled. We still rely on Qemu to
+     * emulate legacy ISA ports, so requires the ISA vga logic.
+     */
+    if (vgt_vga_enabled && pcmc->pci_enabled) {
+        vgt_vga_init(pci_bus);
+        isa_create_simple(isa_bus, "isa-vga");
+    } else {
+        pc_vga_init(isa_bus, pcmc->pci_enabled ? pci_bus : NULL);
+    }
 
     assert(pcms->vmport != ON_OFF_AUTO__MAX);
     if (pcms->vmport == ON_OFF_AUTO_AUTO) {
diff --git a/hw/pci-host/piix.c b/hw/pci-host/piix.c
index f9218aa..91f7ed7 100644
--- a/hw/pci-host/piix.c
+++ b/hw/pci-host/piix.c
@@ -156,8 +156,8 @@ static void i440fx_update_memory_mappings(PCII440FXState *d)
 }
 
 
-static void i440fx_write_config(PCIDevice *dev,
-                                uint32_t address, uint32_t val, int len)
+void i440fx_write_config(PCIDevice *dev,
+                         uint32_t address, uint32_t val, int len)
 {
     PCII440FXState *d = I440FX_PCI_DEVICE(dev);
 
@@ -315,6 +315,10 @@ static void i440fx_realize(PCIDevice *dev, Error **errp)
     if (object_property_get_bool(qdev_get_machine(), "iommu", NULL)) {
         error_report("warning: i440fx doesn't support emulated iommu");
     }
+
+    if (vgt_vga_enabled && xen_enabled()) {
+        vgt_bridge_pci_conf_init(dev);
+    }
 }
 
 PCIBus *i440fx_init(const char *host_type, const char *pci_type,
diff --git a/hw/pci/pci.c b/hw/pci/pci.c
index 7f32edf..4bbc1b6 100644
--- a/hw/pci/pci.c
+++ b/hw/pci/pci.c
@@ -23,6 +23,7 @@
  */
 #include "qemu/osdep.h"
 #include "hw/hw.h"
+#include "hw/i386/pc.h"
 #include "hw/pci/pci.h"
 #include "hw/pci/pci_bridge.h"
 #include "hw/pci/pci_bus.h"
@@ -968,6 +969,11 @@ static PCIDevice *do_pci_register_device(PCIDevice *pci_dev, PCIBus *bus,
     if (devfn < 0) {
         for(devfn = bus->devfn_min ; devfn < ARRAY_SIZE(bus->devices);
             devfn += PCI_FUNC_MAX) {
+
+            /* If vGT/XenGT is in use, reserve 00:02.* for the IGD */
+            if ((vgt_vga_enabled) && devfn == 0x10)
+                continue;
+
             if (!bus->devices[devfn])
                 goto found;
         }
@@ -1002,6 +1008,7 @@ static PCIDevice *do_pci_register_device(PCIDevice *pci_dev, PCIBus *bus,
     pci_dev->irq_state = 0;
     pci_config_alloc(pci_dev);
 
+    printf("set vendor id(%x) for devfn(%x)\n", pc->vendor_id, devfn);
     pci_config_set_vendor_id(pci_dev->config, pc->vendor_id);
     pci_config_set_device_id(pci_dev->config, pc->device_id);
     if (pci_dev->revision_id > 0xff) {
@@ -1047,6 +1054,11 @@ static PCIDevice *do_pci_register_device(PCIDevice *pci_dev, PCIBus *bus,
         config_read = pci_default_read_config;
     if (!config_write)
         config_write = pci_default_write_config;
+    if (vgt_vga_enabled &&
+        config_write == i440fx_write_config) {
+        config_write = vgt_bridge_pci_write;
+    }
+
     pci_dev->config_read = config_read;
     pci_dev->config_write = config_write;
     bus->devices[devfn] = pci_dev;
diff --git a/include/hw/i386/pc.h b/include/hw/i386/pc.h
index 0d9b2c9..8cde9c0 100644
--- a/include/hw/i386/pc.h
+++ b/include/hw/i386/pc.h
@@ -328,6 +328,16 @@ int isa_vga_mm_init(hwaddr vram_base,
                     hwaddr ctrl_base, int it_shift,
                     MemoryRegion *address_space);
 
+/* vgt_vga.c */
+extern int vgt_low_gm_sz;
+extern int vgt_high_gm_sz;
+extern int vgt_fence_sz;
+
+DeviceState *vgt_vga_init(PCIBus *pci_bus);
+void vgt_bridge_pci_conf_init(PCIDevice *dev);
+void vgt_bridge_pci_write(PCIDevice *dev,
+                          uint32_t address, uint32_t val, int len);
+
 /* ne2000.c */
 static inline bool isa_ne2000_init(ISABus *bus, int base, int irq, NICInfo *nd)
 {
diff --git a/include/hw/pci/pci_host.h b/include/hw/pci/pci_host.h
index ba31595..dc7d266 100644
--- a/include/hw/pci/pci_host.h
+++ b/include/hw/pci/pci_host.h
@@ -70,4 +70,7 @@ extern const MemoryRegionOps pci_host_conf_be_ops;
 extern const MemoryRegionOps pci_host_data_le_ops;
 extern const MemoryRegionOps pci_host_data_be_ops;
 
+void i440fx_write_config(PCIDevice *dev,
+                         uint32_t address, uint32_t val, int len);
+
 #endif /* PCI_HOST_H */
diff --git a/include/sysemu/sysemu.h b/include/sysemu/sysemu.h
index 66c6f15..a6ff175 100644
--- a/include/sysemu/sysemu.h
+++ b/include/sysemu/sysemu.h
@@ -133,12 +133,13 @@ extern int autostart;
 
 typedef enum {
     VGA_NONE, VGA_STD, VGA_CIRRUS, VGA_VMWARE, VGA_XENFB, VGA_QXL,
-    VGA_TCX, VGA_CG3, VGA_DEVICE, VGA_VIRTIO,
+    VGA_TCX, VGA_CG3, VGA_DEVICE, VGA_VIRTIO, VGA_VGT,
     VGA_TYPE_MAX,
 } VGAInterfaceType;
 
 extern int vga_interface_type;
 #define xenfb_enabled (vga_interface_type == VGA_XENFB)
+#define vgt_vga_enabled (vga_interface_type == VGA_VGT)
 
 extern int graphic_width;
 extern int graphic_height;
diff --git a/qemu-options.hx b/qemu-options.hx
index c534a2f..8dcf6dc 100644
--- a/qemu-options.hx
+++ b/qemu-options.hx
@@ -1184,7 +1184,7 @@ Rotate graphical output some deg left (only PXA LCD).
 ETEXI
 
 DEF("vga", HAS_ARG, QEMU_OPTION_vga,
-    "-vga [std|cirrus|vmware|qxl|xenfb|tcx|cg3|virtio|none]\n"
+    "-vga [std|cirrus|vmware|qxl|xenfb|tcx|cg3|xengt|none]\n"
     "                select video card type\n", QEMU_ARCH_ALL)
 STEXI
 @item -vga @var{type}
@@ -1219,11 +1219,42 @@ for sun4m machines available in both 1024x768 (OpenBIOS) and 1152x900 (OBP)
 resolutions aimed at people wishing to run older Solaris versions.
 @item virtio
 Virtio VGA card.
+@item xengt
+Intel IGD card based on mediated pass-through technique. A single
+IGD card can be used to accelerate multiple VMs, which each run
+native graphics driver inside.
 @item none
 Disable VGA card.
 @end table
 ETEXI
 
+DEF("vgt_low_gm_sz", 1, QEMU_OPTION_vgt_low_gm_sz,
+    "-vgt_low_gm_sz [MB] allocated aperture size\n",
+    QEMU_ARCH_ALL)
+STEXI
+@item -vgt_low_gm_sz @var{size}
+@findex -vgt_low_gm_sz
+Set the allocated aperture size.
+ETEXI
+
+DEF("vgt_high_gm_sz", 1, QEMU_OPTION_vgt_high_gm_sz,
+    "-vgt_high_gm_sz [MB] allocated GM size\n",
+    QEMU_ARCH_ALL)
+STEXI
+@item -vgt_high_gm_sz @var{size}
+@findex -vgt_high_gm_sz
+Set the allocated aperture size.
+ETEXI
+
+DEF("vgt_fence_sz", 1, QEMU_OPTION_vgt_fence_sz,
+    "-vgt_fence_sz [num] allocated number of fence registers\n",
+    QEMU_ARCH_ALL)
+STEXI
+@item -vgt_fence_sz @var{size}
+@findex -vgt_fence_sz
+Set the allocated aperture size.
+ETEXI
+
 DEF("full-screen", 0, QEMU_OPTION_full_screen,
     "-full-screen    start in full screen\n", QEMU_ARCH_ALL)
 STEXI
diff --git a/vl.c b/vl.c
index 75b788f..9489fd8 100644
--- a/vl.c
+++ b/vl.c
@@ -2088,9 +2088,10 @@ static void select_vgahw(const char *p)
                 error_report("%s not available", ti->name);
                 exit(1);
             }
-            vga_interface_type = t;
-            break;
-        }
+	    vga_interface_type = t;
+	    printf("%s vga is %s\n", __func__, ti->opt_name);           
+	    break;
+	}
     }
     if (t == VGA_TYPE_MAX) {
     invalid_vga:
@@ -2110,6 +2111,7 @@ static void select_vgahw(const char *p)
         } else goto invalid_vga;
         opts = nextopt;
     }
+
 }
 
 typedef enum DisplayType {
@@ -4047,6 +4049,24 @@ int main(int argc, char **argv, char **envp)
                     exit(1);
                 }
                 break;
+            case QEMU_OPTION_vgt_low_gm_sz:
+                {
+                    char *ptr;
+                    vgt_low_gm_sz = strtol(optarg, &ptr, 10);
+                }
+                break;
+            case QEMU_OPTION_vgt_high_gm_sz:
+                {
+                    char *ptr;
+                    vgt_high_gm_sz = strtol(optarg, &ptr, 10);
+                }
+                break;
+            case QEMU_OPTION_vgt_fence_sz:
+                {
+                    char *ptr;
+                    vgt_fence_sz = strtol(optarg, &ptr, 10);
+                }
+                break;
             default:
                 os_parse_cmd_args(popt->index, optarg);
             }
diff --git a/xen-hvm.c b/xen-hvm.c
index a6b404a..2348f81 100644
--- a/xen-hvm.c
+++ b/xen-hvm.c
@@ -199,7 +199,11 @@ static void xen_ram_init(PCMachineState *pcms,
         user_lowmem = HVM_BELOW_4G_RAM_END;
     }
 
-    if (ram_size >= user_lowmem) {
+    if (vgt_vga_enabled &&
+        ram_size >= 0xc0000000) {
+        pcms->above_4g_mem_size = ram_size - 0xc0000000;
+        pcms->below_4g_mem_size = 0xc0000000;
+    } else if (ram_size >= user_lowmem) {
         pcms->above_4g_mem_size = ram_size - user_lowmem;
         pcms->below_4g_mem_size = user_lowmem;
     } else {
